app/scripts/controllers/metametrics.js

```diff
@@ -0,0 +1,409 @@
+    // If we are tracking sensitive data we will always use the anonymousId property
+    // as well as our METAMETRICS_ANONYMOUS_ID. This prevents us from associating potentially
+    // identifiable information with a specific id. During the opt in flow we will track all
+    // events, but do so with the anonymous id. The one exception to that rule is after the
+    // user opts in to MetaMetrics. When that happens we receive back the user's new MetaMetrics
+    // id before it is fully persisted to state. To avoid a race condition we explicitly pass the
+    // new id to the track method. In that case we will track the opt in event to the user's id.
+    // In all other cases we use the metaMetricsId from state.
+    if (excludeMetaMetricsId || (isOptIn && !metaMetricsIdOverride)) {
+      idType = 'anonymousId'
+      idValue = METAMETRICS_ANONYMOUS_ID
+    } else if (isOptIn && metaMetricsIdOverride) {
+      idValue = metaMetricsIdOverride
+    }
+    payload[idType] = idValue
+
+    // Promises will only resolve when the event is sent to segment. For any
+    // event that relies on this promise being fulfilled before performing UI
+    // updates, or otherwise delaying user interaction, supply the 'flushImmediately'
+    // flag to the trackEvent method.
+    return new Promise((resolve, reject) => {
+      const callback = (err) => {
+        if (err) {
+          return reject(err)
+        }
+        return resolve()
+      }
+
+      const target = matomoEvent === true ? segmentLegacy : segment
+
+      target.track(payload, callback)
+      if (flushImmediately) {
+        target.flush()
+      }
+    })
+  }
+
+  trackPage(name, params, environmentType, page, referrer) {
+    const idTrait = this.participateInMetaMetrics ? 'userId' : 'anonymousId'
+    const idValue = this.participateInMetaMetrics
+      ? this.metaMetricsId
+      : METAMETRICS_ANONYMOUS_ID
+    console.log(idTrait, idValue)
+    segment.page(
+      {
+        [idTrait]: idValue,
+        name,
+        properties: {
+          // We do not want to send addresses or accounts in any events
+          // Some routes include these as params.
+          params,
+          locale: this.locale,
+          network: this.network,
+          chain_id: this.chainId,
+          environment_type: environmentType,
+        },
+        context: this._buildContext(referrer, page),
+      },
+      (err, data) => {
+        console.log(err, data)
+      },
+    )
+  }
+
+  /**
+   *
+   * @param {MetaMetricsEventPayload} payload - details of the event
+   * @param {MetaMetricsEventOptions} options - options for handling/routing the event
+   * @returns {Promise<void>}
+   */
+  trackEvent(payload, options) {
+    // event and category are required fields for all payloads
+    if (!payload.event || !payload.category) {
+      throw new Error('Must specify event and category.')
+    }
+
+    if (!this.participateInMetaMetrics && !options.isOptIn) {
+      return Promise.resolve()
```

Comment:
Better to make this function `async` if it's doing async things. Right now it has an inconsistent return value in the event of failure - it'll throw an Error directly in some cases, but it'll return a rejected Promise in others.